### DISCUSSION TITLE
Effective and correct way of invalidating data-cache for arm64 Xen domain

### DISCUSSION THREAD

**Discussion by xakep-amatop** [zephyrproject-rtos/zephyr#58436] [Q&A] [answered]
If we build Zephyr as a Xen Initial domain or for xenvm board, it won't work with cache management support enabled. Initialization of board(s) is failed in [enable_mmu_el1](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm64/core/mmu.c#L805) function during call `sys_cache_data_invd_all` --- this call has set/way instructions. Unfortunately, set/way ops are not easily virtualized by Xen. S/W emulation is disabled, because IP-MMU is active for Dom0. IP-MMU is a IO-MMU made by Renesas, as any good IO-MMU, it shares page-tables with CPU. Trying to emulate S/W with IP-MMU active will lead to IO-MMU faults.

Our team have found a fast workaround for the issue: just replace  [sys_cache_data_invd_all](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm64/core/mmu.c#L819) call with the next one `sys_cache_data_invd_range((void *)CONFIG_KERNEL_VM_BASE, CONFIG_KERNEL_VM_SIZE)`. But this approach may take long time during the initialization of the board --- the larger the virtual memory size we have, the longer the function works.

So, the question is:
**Is there a way to invalidate dcache more effectively without using set/way instructions?**

**Comment by lorc** [comment, marked as answer]
FYI, U-boot had the same issue and there is the fix they implemented: https://www.mail-archive.com/u-boot@lists.denx.de/msg464402.html

**Comment by lorc** [comment]
While this worked well for U-Boot, I am not sure that this is the best approach for Zephyr.

IMO, it is better to clean/invalidate every page during mapping.
